R/drive_download.R

```diff
@@ -43,49 +43,104 @@ drive_download <- function(file = NULL,
       )
-    }
-    mime_type <- drive_mime_type(ext)
-    request <- generate_request(endpoint = "drive.files.export",
-                                params = list(fileId = file$id,
-                                              mimeType = mime_type ))
+    )
   } else {
-    request <- generate_request(endpoint = "drive.files.get",
-                                params = list(alt = "media",
-                                              fileId = file$id))
+    request <- generate_request(
+      endpoint = "drive.files.get",
+      params = list(
+        fileId = file$id,
+        alt = "media"
+      )
+    )
   }
 
-  response <- make_request(request, httr::write_disk(out_path, overwrite = overwrite))
-  success <- file.exists(out_path)
+  response <- make_request(
+    request,
+    httr::write_disk(out_path, overwrite = overwrite)
+  )
+  success <- httr::status_code(response) == 200 && file.exists(out_path)
 
   if (success) {
     if (verbose) {
       message(
-        glue("File downloaded from Google Drive:\n{sq(file$name)}\nSaved locally as:\n{sq(out_path)}")
+        glue(
+          "File downloaded from Google Drive:\n{sq(file$name)}\n",
+          "Saved locally as:\n{sq(out_path)}"
+        )
       )
     }
   } else {
-    spf("Zoinks! the file doesn't seem to have downloaded")
+    stop("Zoinks! the file doesn't seem to have downloaded", call. = FALSE)
   }
   invisible(file)
 }
 
+## get the default export MIME type for a native Google MIME type
+## examples:
+##    Google Doc --> MS Word
+##  Google Sheet --> MS Excel
+## Google Slides --> MS PowerPoint
+get_export_mime_type <- function(mime_type) {
+  m <- .drive$translate_mime_types$mime_type_google == mime_type &
+    is_true(.drive$translate_mime_types$default)
+  if (!any(m)) {
+    stop(glue("Not a recognized Google MIME type:\n{mime_type}"), call. = FALSE)
+  }
+  .drive$translate_mime_types$mime_type_local[m]
+}
+
+## affirm that export_type is a valid export MIME type for a native Google file
+## of type mime_type
+verify_export_mime_type <- function(mime_type, export_type) {
+  m <- .drive$translate_mime_types$mime_type_google == mime_type
+  ok <- export_type %in% .drive$translate_mime_types$mime_type_local[m]
+  if (!ok) {
+    ## to be really nice, we would look these up in drive_mime_type() tibble
+    ## and use the human_below, if found
+    stop(
+      glue(
+        "Cannot export Google file of type:\n{mime_type}\n",
+        "as a file of type:\n{export_type}"
+      ),
+      call. = FALSE
+    )
+  }
+  export_type
+}
+
+## determine file extension from mime_type
+## apply to out_path if not already present
+apply_extension <- function(out_path, mime_type) {
+  mime_tbl <- drive_mime_type()
+  m <- mime_tbl$mime_type == mime_type
+  if (sum(m) != 1) return(out_path)
+
+  ext <- mime_tbl$ext[m]
+  ext_orig <- file_ext_safe(out_path)
+  if (!identical(ext, ext_orig)) {
+    out_path <- paste(out_path, ext, sep = ".")
+  }
+  out_path
+}
```

Comment:
This is a slightly ugly helper. If we keep it, perhaps you should write that function to get extension from MIME type and then use it here. So break this up into two functions: one that gets extension, one that applies it. Move the first into the MIME type file (or path-utils.R?) and the second into path-utils.R. Regardless, this needs some tests.